hydroshare_jupyter_sync/lib/filesystem/fs_map.py

```diff
@@ -0,0 +1,217 @@
+
+        for resource in fs_map._get_resource_ids():
+            # create new instance of LocalFSResourceMap
+            fs_map[resource] = LocalFSResourceMap(fs_map.fs_root / resource)
+
+        return fs_map
+
+    def add_resource(self, resource_id: ResourceId) -> None:
+        """Create new LocalFSResourceMap and add to LocalFSMap instance. `resource_id` provided must
+        be direct child directory of `fs_root`."""
+        if resource_id not in self.data:
+            # create new local resource map
+            r_map = LocalFSResourceMap.from_resource_path(self.fs_root / resource_id)
+
+            # add local resource map to dictionary
+            self.data[resource_id] = r_map
+
+    def add_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        if resource_id in self.data:
+            self.data[resource_id].add_file(relative_resource_file)
+
+    def update_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Update a file in a LocalFSResourceMap instance for a given resource id."""
+        if resource_id in self.data:
+            self.data[resource_id].update_file(relative_resource_file)
+
+    def delete_resource_file(
+        self, resource_id: ResourceId, relative_resource_file: Union[Path, str]
+    ) -> None:
+        """Remove a file in a LocalFSResourceMap instance for a given resource id."""
+        if resource_id in self.data:
+            self.data[resource_id].delete_file(relative_resource_file)
+
+
+class RemoteFSMap(FSMap):
+    """Class representing the relationship between remote HydroShare resource's, resource files, and
+    resource MD5 Hashes."""
+
+    def __init__(
+        self, fs_root: Union[str, Path], hydroshare: HydroShare
+    ) -> Dict[ResourceId, FSResourceMap]:
+        super().__init__()
+        self.fs_root = Path(fs_root).expanduser().resolve()
+        self._hydroshare = hydroshare
+
+    # override
+    @classmethod
+    def create_map(
+        cls, fs_root: Union[Path, str], hydroshare: HydroShare
+    ) -> "RemoteFSMap":
+        # create class instance
+        fs_map = cls(fs_root, hydroshare)
+
+        # NOTE: assumes user if logged in and using standard username, pass auth. Likely should
+        # guard in future.
+        # get username
+        username = hydroshare._hs_session._session.auth[0]
+
+        # NOTE: assumes only resources desired for tracking are owned. may not be desirable.
+        # get user owned resources
+        remote_resources = {
+            res.resource_id for res in hydroshare.search(owner=username)
+        }
+        # naively get local resources based on fs_root location and directory name length
+        naive_local_resources = set(fs_map._get_resource_ids())
+
+        # resources s.t. user is owner and some files from res are local
+        users_local_resources = naive_local_resources & remote_resources
```

Comment:
As you guessed above, there are resources that a user can edit that they aren't the owner of.  There is a "can edit" sharing permission.

Reply:
Okay that is really helpful to know. Thanks for mentioning that. So, just so we are on the same page, using `hsclient.HydroShare.search(edit_permission=True)` is what should be used here?

Reply:
I made changes related to this comment in d2fc2bd.